File: app/models/stats.ts
import type { Statistics, TeamStatistics } from './boxScore';

export const trueShooting = ({
	points,
	fieldGoalsAttempted,
	freeThrowsAttempted,
}: {
	points: number;
	fieldGoalsAttempted: number;
	freeThrowsAttempted: number;
}) => {
	if (fieldGoalsAttempted + freeThrowsAttempted <= 0) {
		return undefined;
	}
	return (0.5 * points) / (fieldGoalsAttempted + 0.475 * freeThrowsAttempted);
};

export const usageRate = (
	{
		fieldGoalsAttempted,
		freeThrowsAttempted,
		turnovers,
		minutes,
	}: {
		fieldGoalsAttempted: number;
		freeThrowsAttempted: number;
		turnovers: number;
		minutes: number;
	},
	teamStats: {
		fieldGoalsAttempted: number;
		freeThrowsAttempted: number;
		turnovers: number;
		minutes: number;
	},
) => {
	return Math.round(
		(100 *
			((fieldGoalsAttempted + 0.44 * freeThrowsAttempted + turnovers) *
				(teamStats.minutes / 5))) /
			(minutes *
				(teamStats.fieldGoalsAttempted +
					0.44 * teamStats.freeThrowsAttempted +
					teamStats.turnovers)),
	);
};

export const pointsPerShot = ({
	points,
	fieldGoalsAttempted,
}: {
	points: number;
	fieldGoalsAttempted: number;
}) => {
	if (fieldGoalsAttempted <= 0) {
		return undefined;
	}
	return Math.round((points * 100) / fieldGoalsAttempted) / 100;
};

export const pie = (
	{
		points,
		fieldGoalsMade,
		freeThrowsMade,
		fieldGoalsAttempted,
		freeThrowsAttempted,
		reboundsDefensive,
		reboundsOffensive,
		assists,
		steals,
		blocks,
		blocksReceived,
		foulsPersonal,
		turnovers,
	}: Statistics,
	teamStats: TeamStatistics,
	otherTeamStats: TeamStatistics,
) => {
	return Math.round(
		100 *
			((points +
				fieldGoalsMade +
				freeThrowsMade -
				fieldGoalsAttempted -
				freeThrowsAttempted +
				reboundsDefensive +
				reboundsOffensive / 2 +
				assists / 2 /* avoid double counting */ +
				steals +
				blocks / 2 -
				blocksReceived / 2 -
				foulsPersonal -
				turnovers) /
				(teamStats.points +
					otherTeamStats.points +
					(teamStats.fieldGoalsMade + otherTeamStats.fieldGoalsMade) +
					(teamStats.freeThrowsMade + otherTeamStats.freeThrowsMade) -
					(teamStats.fieldGoalsAttempted + otherTeamStats.fieldGoalsAttempted) -
					(teamStats.freeThrowsAttempted + otherTeamStats.freeThrowsAttempted) +
					(teamStats.reboundsDefensive + otherTeamStats.reboundsDefensive) +
					(teamStats.reboundsOffensive + otherTeamStats.reboundsOffensive) / 2 +
					(teamStats.assists + otherTeamStats.assists) /
						2 /* avoid double counting */ +
					(teamStats.steals + otherTeamStats.steals) +
					(teamStats.blocks + otherTeamStats.blocks) / 2 -
					(teamStats.blocksReceived + otherTeamStats.blocksReceived) / 2 -
					(teamStats.foulsPersonal + otherTeamStats.foulsPersonal) -
					(teamStats.turnovers + otherTeamStats.turnovers))),
	);
};

File: src/stats.rs
use crate::models::{Statistics, TeamStatistics};

pub fn points_per_shot(stats: &Statistics) -> Option<f64> {
    if stats.field_goals_attempted <= 0 {
        return None;
    }
    Some(((stats.points as f64 * 100.0) / stats.field_goals_attempted as f64).round() / 100.0)
}

pub fn usage_rate(stats: &Statistics, team: &TeamStatistics) -> Option<i64> {
    let minutes = stats.minutes as f64;
    if minutes <= 0.0 {
        return None;
    }

    let numerator = (stats.field_goals_attempted as f64
        + 0.44 * stats.free_throws_attempted as f64
        + stats.turnovers as f64)
        * (team.minutes as f64 / 5.0);
    let denominator = minutes
        * (team.field_goals_attempted as f64
            + 0.44 * team.free_throws_attempted as f64
            + team.turnovers as f64);

    if denominator == 0.0 {
        None
    } else {
        Some((100.0 * numerator / denominator).round() as i64)
    }
}

pub fn pie(stats: &Statistics, team: &TeamStatistics, other: &TeamStatistics) -> i64 {
    let player =
        stats.points as f64 + stats.field_goals_made as f64 + stats.free_throws_made as f64
            - stats.field_goals_attempted as f64
            - stats.free_throws_attempted as f64
            + stats.rebounds_defensive as f64
            + stats.rebounds_offensive as f64 / 2.0
            + stats.assists as f64 / 2.0
            + stats.steals as f64
            + stats.blocks as f64 / 2.0
            - stats.blocks_received as f64 / 2.0
            - stats.fouls_personal as f64
            - stats.turnovers as f64;

    let total = team.points as f64
        + other.points as f64
        + team.field_goals_made as f64
        + other.field_goals_made as f64
        + team.free_throws_made as f64
        + other.free_throws_made as f64
        - team.field_goals_attempted as f64
        - other.field_goals_attempted as f64
        - team.free_throws_attempted as f64
        - other.free_throws_attempted as f64
        + team.rebounds_defensive as f64
        + other.rebounds_defensive as f64
        + (team.rebounds_offensive + other.rebounds_offensive) as f64 / 2.0
        + (team.assists + other.assists) as f64 / 2.0
        + team.steals as f64
        + other.steals as f64
        + (team.blocks + other.blocks) as f64 / 2.0
        - stats.blocks_received as f64 / 2.0
        - team.fouls_personal as f64
        - other.fouls_personal as f64
        - team.turnovers as f64
        - other.turnovers as f64;

    if total == 0.0 {
        0
    } else {
        (100.0 * player / total).round() as i64
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn points_per_shot_handles_zero_attempts() {
        assert_eq!(points_per_shot(&Statistics::default()), None);
    }

    #[test]
    fn points_per_shot_rounds_to_two_decimals() {
        let stats = Statistics {
            points: 25,
            field_goals_attempted: 12,
            ..Statistics::default()
        };
        assert_eq!(points_per_shot(&stats), Some(2.08));
    }
}
